File: scripts/ui/render.py
"""F-043 — single render for an agent chain (sequence + parallel groups).

``scripts/work.py`` and ``scripts/ui/dashboard.py`` previously kept
near-identical 33-line copies of ``_render_agent_chain`` /
``_render_chain``. F-043 centralizes the renderer here.

The function is pure: takes the resolved agent list and the parallel
groups, returns the formatted string. No I/O.
"""

from __future__ import annotations


PARALLEL_TOKEN: str = " ∥ "
SEQUENCE_TOKEN: str = " → "
COMMA_JOIN: str = ", "


def render_agent_chain(
    agents: list[str],
    groups: list[list[str]],
    *,
    parallel_token: str = PARALLEL_TOKEN,
    sequence_token: str = SEQUENCE_TOKEN,
    comma_join: str = COMMA_JOIN,
) -> str:
    """Render an agent chain for both ``work.py`` activate output and the
    dashboard's ``agent chain:`` line.

    Args:
        agents: ordered, deduped agent list (the kickoff routing result).
        groups: list of agent groups that may run in parallel; each group
            is a list of agent names.
        parallel_token: the separator inside a parallel group (default
            ``" ∥ "``).
        sequence_token: the separator between sequential steps (default
            ``" → "``).
        comma_join: fallback joiner used when there are no parallel
            groups (preserves the pre-F-039 zero-diff behavior).

    Returns:
        - ``", ".join(agents)`` when ``groups`` is empty (legacy zero-diff).
        - Otherwise a chain like ``a → (b ∥ c) → d`` collapsing each
          contiguous run that lives inside a declared group into a single
          parenthesized parallel block.
    """
    if not groups:
        return comma_join.join(agents)

    group_sets = [set(g) for g in groups]
    parts: list[str] = []
    i = 0
    while i < len(agents):
        member = agents[i]
        matched = next((gs for gs in group_sets if member in gs), None)
        if matched is None:
            parts.append(member)
            i += 1
            continue
        block: list[str] = []
        while i < len(agents) and agents[i] in matched:
            block.append(agents[i])
            i += 1
        if len(block) >= 2:
            parts.append("(" + parallel_token.join(block).strip() + ")")
        else:
            parts.append(block[0])
    return sequence_token.join(parts).strip()


__all__ = ["render_agent_chain", "PARALLEL_TOKEN", "SEQUENCE_TOKEN", "COMMA_JOIN"]

File: src/ui/render.ts
/**
 * Agent chain renderer — sequence + parallel groups (F-092 port of
 * `scripts/ui/render.py`, originally F-043).
 *
 * Used by both the work activate output and the dashboard's `agent
 * chain:` line. Pure: takes the resolved agent list and the parallel
 * groups, returns the formatted string. No I/O.
 *
 * @module ui/render
 */

/** Default token between parallel-group members. */
export const PARALLEL_TOKEN = ' ∥ ';

/** Default token between sequential steps. */
export const SEQUENCE_TOKEN = ' → ';

/** Fallback joiner used when no parallel groups are declared. */
export const COMMA_JOIN = ', ';

/** Optional input for {@link renderAgentChain}. */
export interface RenderAgentChainOptions {
  parallelToken?: string;
  sequenceToken?: string;
  commaJoin?: string;
}

/**
 * Renders an agent chain.
 *
 * @param agents - Ordered, deduped agent list (kickoff routing
 *   result).
 * @param groups - Parallel-capable groups; each group is a list of
 *   agent names.
 * @returns Either the comma-joined `agents` (legacy zero-diff when
 *   `groups` is empty) or a chain like `a → (b ∥ c) → d` collapsing
 *   contiguous runs that match a group into a parenthesized parallel
 *   block.
 */
export function renderAgentChain(
  agents: ReadonlyArray<string>,
  groups: ReadonlyArray<ReadonlyArray<string>>,
  options: RenderAgentChainOptions = {},
): string {
  const parallelToken = options.parallelToken ?? PARALLEL_TOKEN;
  const sequenceToken = options.sequenceToken ?? SEQUENCE_TOKEN;
  const commaJoin = options.commaJoin ?? COMMA_JOIN;

  if (groups.length === 0) {
    return agents.join(commaJoin);
  }

  const groupSets = groups.map((g) => new Set(g));
  const parts: string[] = [];
  let i = 0;
  while (i < agents.length) {
    const member = agents[i]!;
    const matched = groupSets.find((gs) => gs.has(member)) ?? null;
    if (matched === null) {
      parts.push(member);
      i++;
      continue;
    }
    const block: string[] = [];
    while (i < agents.length && matched.has(agents[i]!)) {
      block.push(agents[i]!);
      i++;
    }
    if (block.length >= 2) {
      parts.push(`(${block.join(parallelToken).trim()})`);
    } else {
      parts.push(block[0]!);
    }
  }
  return parts.join(sequenceToken).trim();
}
